Fix massaction loading and reaction name bug

xpspectre opened the pull request:
In `LoadModelMassAction`, when parsing reaction names, the last few tokens are simply passed into `AddReaction`. In the `Equilibrium.txt` test model, the reaction names are separated by a comma, which is passed in as one of these names. The parser has been fixed to properly take the actual names.

Question: Is the last part of the line `a_ b c 0 kon koff On,Off` in `Equilibrium.txt` supposed to read that way? The tokenizer regex in `LoadModelMassAction` thinks the names are `{'On', ',', 'Off'}`. My fix takes care of this case plus `a_ b c 0 kon koff On Off`.

In `fixReactionName`, the number of names in a cell array of `name` isn't properly parsed. This manifested as the reverse reaction name never being set. The cases have been fixed.

Since rxn names are only used for user organization, it doesn't change any results, but this is definitely a bug.

drhagen commented:
Good find!

I think the comma in `Equilibrium.txt` is left over from a time where the tokenizer didn't split on commas and reverse name was manually extracted by splitting the names by commas in a second step. I must have forgotten to update that file when I changed the behavior. I am inclined keep the current behavior of `LoadModelMassAction` and update `Equilibrium.txt` instead. But if you want to allow commas, I don't really mind.

But `fixReactionName` is clearly a bug.

xpspectre commented:
Changing `Equilibrium.txt` is fine with me. So the spec would be: "fields" 7 and 8 are the forward and reverse reaction names, respectively. They're both optional, or just the forward rxn name can be specified.

It may still be useful to explicitly assign a variable called `names` with fields 7 and 8, and pass that as the 2nd arg into `AddReaction`.

I'm also fine with leaving it as is. I don't have any issues with backwards compatibility, but others may. Is this officially documented anywhere? Either way, it seems like it's changed.

xpspectre commented:
Actually, the tokenizer line is this: `tokens = vec(regexp(line, ',|".*"|[^\s,]*','match'));` The last comma (inside `[^\s,]`) allows `line` to break up along commas, but the comma also becomes a token.

Making the user specify everything whitespace-delimited means the tokenizer can become `tokens = vec(regexp(line, '".*"|[^\s]*','match'));`.

drhagen commented:
> It may still be useful to explicitly assign a variable called `names` with fields 7 and 8, and pass that as the 2nd arg into `AddReaction`.

This sounds good to me.

> The last comma (inside `[^\s,]`) allows `line` to break up along commas, but the comma also becomes a token.

There are several places where commas have syntactic meaning in the file definition. There is a question about what to do on lines where commas don't have meaning. It looks like I tried to make the tokenizer consistent (though looking at the file, I missed the states section, which uses a different tokenizer). I think we should make the tokenizer consistent across the file, which means commas are treated specially. I am open to alternatives, though. This is not documented anywhere. Even the Biolucia parser has better documentation at this point.

xpspectre commented:
Well, the Biolucia parser uses a formal grammar so it's definitely specified.

I don't see any example models that have a syntatically significant comma (or insignificant, i.e., inside a quoted name). They should be in `Tutorial/Testing.txt` or something in `Testing/`, but I don't see any. 

I recall that states whose initial condition is a linear combo of seeds has commas separating the seeds, and outputs whose definition is a linear combo of states is similar. Anything else?

drhagen commented:
> I don't see any example models that have a syntatically significant comma (or insignificant, i.e., inside a quoted name). They should be in `Tutorial/Testing.txt` or something in `Testing/`, but I don't see any.
>
> I recall that states whose initial condition is a linear combo of seeds has commas separating the seeds, and outputs whose definition is a linear combo of states is similar. 

Well, that's a problem if there are no tests for this; it is a pretty complicated feature. (I wonder if it still even works correctly.) You are correct that the commas are for defining linear combinations in states and outputs.

xpspectre commented:
This updated commit resolves to eliminate syntactic commas altogether, replacing them with whitespace everywhere. The tokenizing expression is now everywhere `tokens = vec(regexp(line, '".*"|[^\s]*','match'));`. This affects

- Linear combos of seeds for state ICs.
- Linear combos of states and inputs for outputs
- Forward and reverse rxn names

Anything missing?

This behavior worked properly by default - did we forget changing it to this?

The model that demonstrates all the massaction model file format features in `Testing/Tutorial.txt` has been updated with these and also added to the unit tests.

I'll squash these commits together into 1 when merging this pull request.

drhagen commented:
This is a lot simpler than the previous behavior. I like it. As this is a breaking change, I would like @apgar's opinion. He's the main user who might be affected by this.

There are a few printed commas in `SaveModel` that need to be changed to printed spaces to be compatible with this change. But I opened that file and found that is out of date anyway. I made a separate issue #94 in case you don't want to deal with that in this pull request.

While looking to see if you updated the compartments section, I also noticed that `LoadModelMassAction` does not read in multiple compartment sizes at all. (This is definitely my fault.) They should syntactically identical to output contributors. That might be a nice one to also fix in this in this pull request.

```
% Compartments
cell 3 10 membrane_proteins=3.5
```

xpspectre commented:
For the more complete compartments section, the 4th and later tokens make up the compartment value, overriding the 3rd token, right?

drhagen commented:
No, the 3rd and later tokens are exactly like the 2nd and later tokens of outputs. Each one is a linear contribution to to the compartment size. For example, the example above represents a compartment value of `10 + membrane_proteins*3.5`. The actual value passed to `AddReaction` would be `{'', 10; 'membrane_proteins', 3.5}`

xpspectre commented:
OK that makes sense. Then it's not actually a linear function but rather an affine one. In that case, it looks like compartments and states allow this, but outputs currently don't. In `FinalizeModel`, the output names are qualified (compartments added if needed). It's assumed that each row of that n x 2 cell matrix has a state or input name in the 1st col. This breaks when there isn't a name, such as the following line in a model

    y1 x1=1.5 x2=2.5 3.5

 There haven't been any tests that would have caught this.

xpspectre commented:
This pull request is seeing some serious scope creep - I just fixed all the comment handling in the last commit. I can break this up into 3 separate pull requests:

1. Fix reaction name bug
2. Fix comment handling
3. Unify linear combos code, disallowing syntactically significant commas, and unifying tokenizer.

drhagen commented:
> This pull request is seeing some serious scope creep - I just fixed all the comment handling in the last commit. I can break this up into 3 separate pull requests

This pull request has become a "fix LoadModelMassAction" pull request. If you want to break things up, feel free, but I don't think it is necessary.

xpspectre commented:
Whoever merges this pull request should squash the commits together with this overall commit message:

Fix massaction loading and reaction name bugs
- Note that this breaks backwards compatibility for some massaction models specified with kroneckerbio's massaction model file format from a long time ago.
- Fixed reaction name parsing bug. In `fixReactionName`, the number of names in a cell array of `name` isn't properly parsed. This manifested as the reverse reaction name never being set. The `Equilibrium.txt` test model has been fixed.
- In `LoadModelMassAction`, the tokenizer and parser for kroneckerbio's massaction model file format have been updated and cleaned up. Specification of:
  - Compartment size from states/inputs
  - State initial condition from seeds
  - Output from states/inputs
as linear combinations (actually affine) has been unified and updated to the current spec (space separated name[=coefficient] terms, no more syntatically significant commas).
- Comments in kroneckerbio's massaction model file format have been fully implemented.
- Test models updated with current usage.

drhagen commented:
I was going through and checking the regexes of `LoadModelMassAction`. There are a few remaining questions:

* Should species names be allowed to contain spaces if they are within double quotes (I am inclined to say yes)
* If they are are allowed to be double quoted, can species names be empty (like "") (I am inclined to say no, mainly because an empty string has special meaning in a lot of places in Kronecker)
* There are still some places where commas hold special meaning (they are not allowed in species names for example). Should this continue to be disallowed or should commas become fully normal non-whitespace?
  * One difficulty is that the secret workaround for multiple-product reactions is to start with a comma

xpspectre commented:
I think from a bunch of issues ago, we settled on the following rules for names of stuff:

1. The name is either a valid Matlab identifier (starts with a letter (lowercase or uppercase) and is followed by any combo of letters, numbers and underscores) w/o double-quotes, OR
2. The name is double-quoted and can contain any string except for the dot ".", which is specially recognized as "compartment.state" or "compartment.input".

Following these rules, points 1 and 3 are clear - both spaces and commas should be allowed inside double-quotes. Commas shouldn't have any special meaning. I'm not aware of any secret usage of multiple-product reactions; is this what the add large reaction function is for?

An empty "" seems to violate the second point - there's no string inside "". Using "" would probably break the code, especially on isempty checks. I think this applies when there's no product or reactant, and probably other places. If a user wants to use a blank as an identifier, it's their own fault. After all, there are other less dumb things the user can do that aren't properly error-checked.

drhagen commented:
I think we are on the same page, @xpspectre. That sounds all reasonable. I changed it to allow commas inside of species names and disallowed empty names entirely. I left the large reaction stuff as is; we can deal with that when we deal with #94.

I rebased this on top of the #95 and the tests pass. If there are no objections, I'll merge tomorrow.

apgar commented:
That works for me.  Have we captured anywhere the full set of
specifications for the mass action file format?  We are working on some
non-matlab based tools and are building translators from kroecker and it
would be good to know we cover all of the cases.

Kind Regards,
Josh

_________________________________________________________________________________


*Joshua Apgar**Applied BioMath | *Co-founder

*Email: apgar@appliedbiomath.com <apgar@appliedbiomath.com>*
*Phone: 617.914.8890*
*Web:   www.appliedbiomath.com <http://www.appliedbiomath.com>*



On Mon, Aug 8, 2016 at 9:17 PM, David Hagen <notifications@github.com>
wrote:

> I think we are on the same page, @xpspectre <https://github.com/xpspectre>.
> That sounds all reasonable. I changed it to allow commas inside of species
> names and disallowed empty names entirely. I left the large reaction stuff
> as is; we can deal with that when we deal with #94
> <https://github.com/kroneckerbio/kroneckerbio/issues/94>.
>
> I rebased this on top of the #95
> <https://github.com/kroneckerbio/kroneckerbio/pull/95> and the tests
> pass. If there are no objections, I'll merge tomorrow.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kroneckerbio/kroneckerbio/pull/93#issuecomment-238426321>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGAyHpYk0NHOw8IR9coakF3ynf3vBWPGks5qd9U4gaJpZM4JUPCC>
> .
>

drhagen commented:
> Have we captured anywhere the full set of specifications for the mass action file format?

No. The right place to write that up would probably be on the Github wiki. The difficulty is that Matlab does not have a way to write a grammar-based parser, so the spec would be totally independent of the implementation. I think we should write up a spec anyway, if nothing else than to serve as something to define when the parser is right or wrong.

xpspectre commented:
I started writing a formal grammar for the massaction file format on the wiki (https://github.com/kroneckerbio/kroneckerbio/wiki/KroneckerBio-Inputs-Spec). But it's really convoluted - well, the first draft anyway. The massaction file format isn't particularly well suited for EBNF grammars because it's basically a bag of different context-specific rules. You basically have to write a new section for compartments, states, inputs, etc. Writing a general parser (I'm actually messing with porting funcparserlib from Python to Matlab) and implementing the rules is at least as complex and error-prone as what we already have.

However, writing a document w/o a formal grammar, and just specifying what's expected for each part of the model, is a good idea.
